### PR TITLE
Add RatingIssue table, link contentions

### PR DESCRIPTION
**Pull request by pkarman**
connects #7178

### Description

* Add `rating_issues` table, re-use existing model class
* Sync `Veteran` rating issues based on matching `contention_reference_id`
* make `contention_reference_id` a unique integer

### Acceptance Criteria
- [x] Code compiles correctly
- [x] Tests continue to pass

### Testing Plan
See the specific UAT examples in #7178

### CODE REVIEW

File: app/models/veteran.rb [hunk 1]
```diff
@@ -137,6 +137,19 @@ def participant_id
     super || ptcpnt_id
   end
 
+  def sync_rating_issues!
+    timely_ratings(from_date: Time.zone.today).each do |rating|
+      rating.issues.each do |rating_issue|
+        next unless rating_issue.contention_reference_id
```
Comment (shanear) [inline app/models/veteran.rb]: If you prefer a more functional style, this can be captured on line `142` by doing something like

```
rating_issues.select(&:contention_reference_id).each do
  # ...
end
```

File: app/models/claim_review.rb [hunk 2]
```diff
@@ -95,6 +96,10 @@ def sync_dispositions(reference_id)
     end
   end
 
+  def sync_rating_issues(veteran_file_number)
+    Veteran.find_by!(file_number: veteran_file_number).sync_rating_issues!
```
Comment (leikkisa) [inline app/models/claim_review.rb]: You should be able to access "veteran" directly since it's a method in ama_review.
```veteran.sync_rating_issues!```

File: app/models/claim_review.rb [hunk 3]
```diff
@@ -71,6 +71,7 @@ def invalid_modifiers
   def on_sync(end_product_establishment)
     if end_product_establishment.status_cleared?
       sync_dispositions(end_product_establishment.reference_id)
+      sync_rating_issues(end_product_establishment.veteran_file_number)
```
Comment (leikkisa) [inline app/models/claim_review.rb]: See comment below, it seems like you could put veteran.sync_rating_issues! here directly.

File: app/models/veteran.rb (hunk 1 above)
Comment (leikkisa) [reply app/models/veteran.rb]: I'm wondering if it's a bad idea for the from_date to be Time.zone.today in case there is a delay between when the award is promulgated, then the end_product cleared, and then when we succeed at syncing.  This searching ratings within a very narrow band of just today (from today, to today).

File: app/services/external_api/bgs_service.rb [hunk 4]
```diff
@@ -257,13 +257,14 @@ def init_client
...
       application: "CASEFLOW",
-      client_ip: Rails.application.secrets.user_ip_address,
+      client_ip: ENV.fetch("USER_IP_ADDRESS", Rails.application.secrets.user_ip_address),
       client_station_id: current_user.station_id,
       client_username: current_user.css_id,
       ssl_cert_key_file: ENV["BGS_KEY_LOCATION"],
       ssl_cert_file: ENV["BGS_CERT_LOCATION"],
       ssl_ca_cert: ENV["BGS_CA_CERT_LOCATION"],
       forward_proxy_url: forward_proxy_url,
+      jumpbox_url: ENV["RUBY_BGS_JUMPBOX_URL"],
```
Comment (leikkisa) [inline app/services/external_api/bgs_service.rb]: I'm curious what this is here for. If there's a forward_proxy_url I don't think the jumpbox_url does anything.

File: db/schema.rb [hunk 5]
```diff
@@ -23,8 +23,8 @@
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
     t.boolean "granted"
-    t.index ["person_id"], name: "index_advance_on_docket_grants_on_person_id"
-    t.index ["user_id"], name: "index_advance_on_docket_grants_on_user_id"
+    t.index ["person_id"], name: "index_advance_on_docket_motions_on_person_id"
```
Comment (leikkisa) [inline db/schema.rb]: These seem like they might be from another PR?

File: app/models/veteran.rb [hunk 6]
```diff
@@ -137,6 +137,19 @@ def participant_id
...
+        request_issue = RequestIssue.find_by(contention_reference_id: rating_issue.contention_reference_id)
+        if request_issue
+          rating_issue.request_issue = request_issue
+          rating_issue.save!
```
Comment (leikkisa) [inline app/models/veteran.rb]: 👍

File: app/services/external_api/bgs_service.rb (hunk 4 above)
Comment (pkarman) [reply app/services/external_api/bgs_service.rb]: I was using `jumpbox_url` to tunnel through the UAT jumpbox in order to test from the rails console, in a similar way to https://github.com/department-of-veterans-affairs/appeals-design-research/blob/master/Projects/Intake/onboarding/testing-bgs.md

File: db/schema.rb (hunk 5 above)
Comment (pkarman) [reply db/schema.rb]: Yes the table name changed in a previous migration but was not fixed in the schema. When I ran `rake db:migrate` locally it re-created the indexes based on the new table name.

**Comment by pkarman** [comment]
@shanear @leikkisa you both made comments about `from_date` being `today...today`

I read the code otherwise: https://github.com/department-of-veterans-affairs/caseflow/pull/7348#discussion_r224493911

If I am reading the code correctly, it seems like `from_date` is not a great param name since you both misinterpreted it. Maybe we should rename it `end_date`?

File: app/models/rating_issue.rb [hunk 7]
```diff
@@ -1,7 +1,14 @@
...
+class RatingIssue < ApplicationRecord
+  belongs_to :request_issue
 
-  attr_accessor :reference_id, :decision_text
+  # this local attribute is used to resolve the related RequestIssue
+  attr_accessor :contention_reference_id
+
+  def save_with_request_issue!
+    return unless related_request_issue
+    self.request_issue = related_request_issue
```
Comment (pkarman) [inline app/models/rating_issue.rb]: we use `save` rather than `update` because Rating uses `new` rather than `create`. I.e. the object may not have been previously `insert`ed.

**Comment by shanear** [comment]
> If I am reading the code correctly, it seems like from_date is not a great param name since you both misinterpreted it. Maybe we should rename it end_date?

I agree, but timely_ratings will be going away soon anyways, so I'm not sure we need to fix it: https://github.com/department-of-veterans-affairs/caseflow/issues/7299

**Comment by pkarman** [comment]
@shanear @leikkisa I removed the WIP on this one -- care to give it an official review?

File: app/models/end_product_establishment.rb [hunk 8]
```diff
@@ -310,7 +310,7 @@ def create_contentions_in_vbms(descriptions)
   end
 
   def contention_for_object(for_object)
-    contentions.find { |contention| contention.id == for_object.contention_reference_id }
+    contentions.find { |contention| contention.id.to_i == for_object.contention_reference_id.to_i }
```
[the remark this answers is not in the document]
Comment (pkarman) [reply app/models/end_product_establishment.rb]: no, unfortunately. all the SOAP responses are parsed as strings, so `contention.id` is a string.

File: spec/models/end_product_establishment_spec.rb [hunk 9]
```diff
@@ -370,19 +370,20 @@
     end
 
     let(:reference_id) { "stevenasmith" }
+    let(:contention_ref_id) { SecureRandom.random_number(1_000_000) }
```
[the remark this answers is not in the document]
Comment (pkarman) [reply spec/models/end_product_establishment_spec.rb]: I was trying to mock what the fakes lib does, but I see your point in this particular test.
